### PR TITLE
Add code cleanup rules (+perf)

### PR DESCRIPTION
**Pull request by BrennanConroy**
[IDE0020: Use pattern matching to avoid is check followed by a cast (with variable)](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0020-ide0038)
[IDE0029: Use coalesce expression (non-nullable types)](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0029-ide0030)
[IDE0030: Use coalesce expression (nullable types)](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0029-ide0030)
[IDE0031: Use null propagation](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0031)
[IDE0038: Use pattern matching to avoid is check followed by a cast (without variable)](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0020-ide0038)
[IDE0200: Lambda expression can be removed](https://github.com/dotnet/roslyn/blob/main/src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryLambdaExpression/CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer.cs)

These rules all can result in improved performance, details explained in https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/
Additionally, it's often cleaner looking code.

### CODE REVIEW

File: src/Http/Routing/src/Matching/DfaMatcherFactory.cs [hunk 1]
```diff
@@ -32,9 +32,6 @@ public override Matcher CreateMatcher(EndpointDataSource dataSource)
         // when the services are disposed.
         var lifetime = _services.GetRequiredService<DataSourceDependentMatcher.Lifetime>();
 
-        return new DataSourceDependentMatcher(dataSource, lifetime, () =>
-        {
-            return _services.GetRequiredService<DfaMatcherBuilder>();
-        });
+        return new DataSourceDependentMatcher(dataSource, lifetime, _services.GetRequiredService<DfaMatcherBuilder>);
```
Comment (davidfowl) [inline src/Http/Routing/src/Matching/DfaMatcherFactory.cs:35]: Is this right?

Comment (BrennanConroy) [reply src/Http/Routing/src/Matching/DfaMatcherFactory.cs:35]: Yeah, it's `_services.GetRequiredService<DfaMatcherBuilder>` not `_services.GetRequiredService<DfaMatcherBuilder>()`.

Comment (davidfowl) [reply src/Http/Routing/src/Matching/DfaMatcherFactory.cs:35]: ahhhh

**Review by Tratcher — APPROVED** [review]

**Pull request merged**